src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs

```diff
@@ -100,9 +102,41 @@ public static class TensorflowCatalog
         /// <param name="options">The <see cref="TensorFlowEstimator.Options"/> specifying the inputs and the settings of the <see cref="TensorFlowEstimator"/>.</param>
         /// <param name="tensorFlowModel">The pre-loaded TensorFlow model.</param>
-        public static TensorFlowEstimator TensorFlow(this TransformsCatalog catalog,
+        public static TensorFlowEstimator TensorFlow(this TransformsCatalog.TensorFlowTransforms catalog,
             TensorFlowEstimator.Options options,
             TensorFlowModelInfo tensorFlowModel)
             => new TensorFlowEstimator(CatalogUtils.GetEnvironment(catalog), options, tensorFlowModel);
+
+        /// <summary>
+        /// This method retrieves the information about the graph nodes of a TensorFlow model as an <see cref="DataViewSchema"/>.
+        /// For every node in the graph that has an output type that is compatible with the types supported by
+        /// <see cref="TensorFlowEstimator"/>, the output schema contains a column with the name of that node, and the
+        /// type of its output (including the item type and the shape, if it is known). Every column also contains metadata
+        /// of kind <see cref="TensorFlowUtils.TensorflowOperatorTypeKind"/>, indicating the operation type of the node, and if that node has inputs in the graph,
+        /// it contains metadata of kind <see cref="TensorFlowUtils.TensorflowUpstreamOperatorsKind"/>, indicating the names of the input nodes.
+        /// </summary>
+        /// <param name="catalog">The transform's catalog.</param>
+        /// <param name="tensorFlowModel">The pre-loaded TensorFlow model. Please see <see cref="LoadTensorFlowModel(TransformsCatalog.TensorFlowTransforms, string)"/> to know more about loading model into memory.</param>
+        public static DataViewSchema GetModelSchema(this TransformsCatalog.TensorFlowTransforms catalog, TensorFlowModelInfo tensorFlowModel)
+            => TensorFlowUtils.GetModelSchema(CatalogUtils.GetEnvironment(catalog), tensorFlowModel.Session.Graph);
+
+        /// <summary>
+        /// This method retrieves the information about the input graph nodes of a TensorFlow model as an <see cref="DataViewSchema"/>.
+        /// The nodes with OpType as "Placeholder" are classified as input nodes. This is the convenience method to inspect only input nodes.
+        /// For retrieving complete information please see <see cref="GetModelSchema(TransformsCatalog.TensorFlowTransforms, TensorFlowModelInfo)"/>
+        /// </summary>
+        /// <param name="catalog">The transform's catalog.</param>
+        /// <param name="tensorFlowModel">The pre-loaded TensorFlow model. Please see <see cref="LoadTensorFlowModel(TransformsCatalog.TensorFlowTransforms, string)"/> to know more about loading model into memory.</param>
+        public static DataViewSchema GetInputSchema(this TransformsCatalog.TensorFlowTransforms catalog, TensorFlowModelInfo tensorFlowModel)
```

Comment:
>GetInputSchema [](start = 37, length = 14)

This API does not seem like any C# API I've seen. It seems more like a C API with a bunch of static functions that you call with an object, rather than anything like what I'd expect in an actual object oriented property. Why is not *everything* you've added as an extension method, except possibly the creation of what you call now this `TensorFlowModelInfo`, just a property *or* a method of that `TensorFlowModelInfo`?

Reply:
Tom, I was under the impression that we wanted to have as much of the functionality as possible accessible from MLContext, and expose as few classes as possible, that's why I suggested exposing these methods here. Do you think the `LoadTensorFlowModel` API should still be an extension method on `TransformsCatalog`, or should we expose the `TensorFlowUtils` class for that?

---
In reply to: [258769868](https://github.com/dotnet/machinelearning/pull/2672#discussion_r258769868) [](ancestors = 258769868)

Reply:
I don't think that's how we've been doing things at all. We use MLContext for things like component creation, but once you, to give the most conspicuous example, create an estimator, the creation of a transformer, the getting of information *out* of the transformer, the usage of the transformer to transform a data view, *that* does not involve the MLContext.

You create a model. The model is the object, you use objects by calling methods and properties on those objects, just like everything else. Recall the title of #1098 that is I'd say the central MLContext... one MLContext to *create* them all, not to *use* them all.

I gave examples above about how we don't structure our APIs as they are structured here, and I think my reasoning is pretty solid.

Reply:
Thanks @TomFinley, I hope I addressed your concern here and other places. I am changing it resolved for now.

---
In reply to: [259531678](https://github.com/dotnet/machinelearning/pull/2672#discussion_r259531678) [](ancestors = 259531678)

Reply:
Marking unresolved... I do not see that anything has changed at all. `TensorFlowModel` has no meaningful methods on it, everything is still being exposed via these extension methods on top of properties.

---
In reply to: [259545815](https://github.com/dotnet/machinelearning/pull/2672#discussion_r259545815) [](ancestors = 259545815,259531678)

Reply:
Thanks Tom for feedback. I have refactored the code according to the your comments.

---
In reply to: [260027159](https://github.com/dotnet/machinelearning/pull/2672#discussion_r260027159) [](ancestors = 260027159,259545815,259531678)